crates/wiggle/src/lib.rs

```diff
@@ -613,14 +623,19 @@ impl<'a, T> GuestPtr<'a, [T]> {
     where
         T: GuestTypeTransparent<'a> + Copy + 'a,
     {
-        // bounds check ...
-        let mut self_slice = self.as_slice_mut()?;
+        // Retrieve the slice of memory to copy to, performing the necessary
+        // bounds checks ...
+        let guest_slice = self.as_unsafe_slice_mut()?;
         // ... length check ...
-        if self_slice.len() != slice.len() {
+        if guest_slice.len != slice.len() {
             return Err(GuestError::SliceLengthsDiffer);
         }
-        // ... and copy!
-        self_slice.copy_from_slice(slice);
+        // ... and copy the bytes. SAFETY:  we copy the same regardless of
+        // whether the memory is shared or non-shared and accept that the guest
+        // data may be concurrently modified. TODO: audit that the underlying
+        // use of `std::ptr::copy` is safe with shared memory
+        // (https://github.com/bytecodealliance/wasmtime/issues/4203)
+        unsafe { std::ptr::copy(slice.as_ptr(), guest_slice.ptr, guest_slice.len) };
```

Comment:
I have `UnsafeGuestSlice::mut_borrow(self)` which then makes it difficult to either use the mutable borrow or use the unsafe slice for `std::ptr::copy`. Should it take `&mut self` instead?

Reply:
While no longer compatible with `?` how about:

```rust
enum BorrowResult<T,S> {
    Ok(T),
    Shared(S),
    Err(GuestError),
}
```

as the return value? So that way given a `Shared` error you  can still get at the internal pointer.

While you could change `mut_burrow` to `&mut self` it would require still having the `self` version to get the lifetimes on `as_slice_mut` to work.